rust_snuba/src/processors/spans.rs

```diff
@@ -0,0 +1,222 @@
+    #[serde(default)]
+    system: Option<String>,
+    #[serde(default)]
+    status: Option<String>,
+    #[serde(default)]
+    status_code: Option<i32>,
+    #[serde(default)]
+    transaction: Option<String>,
+    #[serde(default, rename(serialize = "transaction.op"))]
+    transaction_op: Option<String>,
+    #[serde(default)]
+    op: Option<String>,
+    #[serde(default, rename(serialize = "transaction.method"))]
+    transaction_method: Option<String>,
+    #[serde(flatten)]
+    extra: BTreeMap<String, Value>,
+}
+
+fn deserialize_optional_uuid<'de, D>(deserializer: D) -> Result<Option<String>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let s: Option<String> = Option::deserialize(deserializer)?;
+    if let Some(s) = s {
+        return Ok(Some(
+            Uuid::parse_str(&s)
+                .map_err(serde::de::Error::custom)?
+                .to_string(),
+        ));
+    }
+    Ok(None)
+}
+
+fn deserialize_uuid<'de, D>(deserializer: D) -> Result<String, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let buf = String::deserialize(deserializer)?;
+    Ok(Uuid::parse_str(&buf)
+        .map_err(serde::de::Error::custom)?
+        .to_string())
+}
+
+fn enforce_retention(_: &SpanMessage) -> Option<u32> {
+    Some(90)
+}
+
+fn process_span_meta(event: &SpanMessage, processed: &mut BTreeMap<&str, Value>) {
+    processed.insert("trace_id", json!(event.trace_id));
+    processed.insert(
+        "span_id",
+        json!(u128::from_str_radix(event.span_id.as_str(), 16).unwrap()),
+    );
+    processed.insert("segment_id", processed.get("span_id").unwrap().clone());
+    processed.insert("is_segment", json!(event.is_segment));
+    processed.insert(
+        "parent_span_id",
+        json!(event.parent_span_id.as_ref().map_or(None, |id| Some(
+            u128::from_str_radix(id.as_str(), 16).unwrap()
+        ))),
+    );
+    processed.insert("transaction_id", json!(event.event_id));
+
+    processed.insert(
+        "description",
+        json!(event.description.as_ref().map_or("", |desc| desc.as_str())),
+    );
+    processed.insert(
+        "group_raw",
+        json!(event
+            .group_raw
+            .as_ref()
+            .map_or(0, |group_raw| u64::from_str_radix(group_raw, 16).unwrap())),
+    );
+}
+
+pub fn process_message(
+    payload: KafkaPayload,
+    _metadata: KafkaMessageMetadata,
+) -> Result<BytesInsertBatch, InvalidMessage> {
+    if let Some(payload_bytes) = payload.payload {
+        let event: SpanMessage = serde_json::from_slice(&payload_bytes).map_err(|err| {
+            let s = err.to_string();
+            log::error!("Failed to deserialize message: {}, {}", err, s);
+            InvalidMessage
+        })?;
+        let retention_days = enforce_retention(&event).ok_or_else(|| {
+            log::error!("Failed to process message: event too old");
+            InvalidMessage
+        })?;
+
+        let mut processed: BTreeMap<&str, Value> = BTreeMap::from([
```

Comment:
instead of btreemap I would create a new struct, similar to how querylog has FromQuerylog for input data and a Querylog for output data

Reply:
I see, I'm assuming you mean something like
``` rust

#[derive(Debug, Deserialize)]
struct Span {
...
}

impl TryFrom<SpanMessage> for Span {
    fn try_from(from: SpanMessage) -> Result<QueryList, InvalidMessage> {
      ...
    }
}
```

Reply:
yup!